📖 update MachineSet PreflightChecks documentation (#8775)
kubernetes-sigs/cluster-api: docs/book/src/tasks/experimental-features/machineset-preflight-checks.md

# Experimental Feature: MachineSetPreflightChecks (alpha)

The `MachineSetPreflightChecks` feature can provide additional safety while creating new Machines and remediating existing unhealthy Machines of a MachineSet.

**Feature gate name**: `MachineSetPreflightChecks`

**Variable name to enable/disable the feature gate**: `EXP_MACHINE_SET_PREFLIGHT_CHECKS`

## Supported PreflightChecks

> 

Reviewer: I think Fabrizio was also mentioning that we should explain why each of those checks is important (aka a short concise description for each what problems might occur if they are not run / skipped)

Reviewer: Not sure if we need a description per preflight check as the problems are similar:
* we would create Machines which versions outside of the official supported version range => it's unclear if the Machine creation woudl work => can lead to massive Machine churn
* control plane stable => as long as the control plane is not stable it's not possible to ensure the join works as it's unclear against which kube-apiserver version we join
* ... maybe more

Author: I added a small block at the top of the page to provide context on the MachineSet Preflight checks (pulled form the ms preflight checks issue). 

PTAL.

# Experimental Feature: MachineSetPreflightChecks (alpha)

The `MachineSetPreflightChecks` feature can provide additional safety while creating new Machines and remediating existing unhealthy Machines of a MachineSet.

When a MachineSet creates machines under certain circumstances, the operation fails or leads to a new machine that will be deleted and recreated in a short timeframe,
leading to unwanted Machine churn. Some of these circumstances include, but not limited to, creating a new Machine when Kubernetes version skew could be violated or 

> joining a Machine when the Control Plane is upgrading leading to failure because of unclear kube-apiserver version. 

Reviewer 2: ```suggestion
joining a Machine when the Control Plane is upgrading leading to failure because of mixed kube-apiserver version or due to the cluster load balancer delays in adapting to the changes. 
```

Enabling `MachineSetPreflightChecks` provides safety in such circumstances by making sure that a Machine is only created when it is safe to do so.

**Feature gate name**: `MachineSetPreflightChecks`

**Variable name to enable/disable the feature gate**: `EXP_MACHINE_SET_PREFLIGHT_CHECKS`

## Supported PreflightChecks

### `ControlPlaneIsStable`

* This preflight check ensures that the ControlPlane is currently stable i.e. the ControlPlane is currently neither provisioning nor upgrading.   
* This preflight check is only performed if:
  * The Cluster uses a ControlPlane provider.

>   * ControlPlane version is defined (`controlPlane.spec.version` is set).

Reviewer: Sorry my previous comment was a bit misleading. What I was trying to say is let's use the path "as the user sees it".
The field paths are fine now but I would continue to write the CR name the way users write it in YAML
`ControlPlane.spec.version`
`MachineSet.spec.template.spec.version`

(ControlPlane obviously doesn't exist but I would still go with the normal naming convention, cf. KubeadmControlPlane)

Feel free to squash after this one

### `KubernetesVersionSkew`

* This preflight check ensures that the MachineSet and the ControlPlane conform to the [Kubernetes version skew](https://kubernetes.io/releases/version-skew-policy/#kubelet).
* This preflight check is only performed if:
    * The Cluster uses a ControlPlane provider.
    * ControlPlane version is defined (`ControlPlane.spec.version` is set).
    * MachineSet version is defined (`MachineSet.spec.template.spec.version` is set).

### `KubeadmVersionSkew`

* This preflight check ensures that the MachineSet and the ControlPlane conform to the [kubeadm version skew](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/#kubeadm-s-skew-against-kubeadm).
* This preflight check is only performed if:
  * The Cluster uses a ControlPlane provider.
  * ControlPlane version is defined (`ControlPlane.spec.version` is set).
  * MachineSet version is defined (`MachineSet.spec.template.spec.version` is set).
  * MachineSet uses the `Kubeadm` Bootstrap provider.

> 

Reviewer 2: ```suggestion
```

## Opting out of PreflightChecks

Once the feature flag is enabled the preflight checks are enabled for all the MachineSets including new and existing MachineSets.
It is possible to opt-out of one or all of the preflight checks on a per MachineSet basis by specifying a comma-separated list of the preflight checks on the 
`machineset.cluster.x-k8s.io/skip-preflight-checks` annotation on the MachineSet.  

Examples: 
* To opt out of all the preflight checks set the `machineset.cluster.x-k8s.io/skip-preflight-checks: All` annotation.
* To opt out of the `ControlPlaneIsStable` preflight check set the `machineset.cluster.x-k8s.io/skip-preflight-checks: ControlPlaneIsStable` annotation.
* To opt out of multiple preflight checks set the `machineset.cluster.x-k8s.io/skip-preflight-checks: ControlPlaneIsStable,KubernetesVersionSkew` annotation.

<aside class="note">

<h1>Pro-tip: Set annotation through MachineDeployment</h1>

Because of the [metadata propagation](../../developer/architecture/controllers/metadata-propagation.md#machinedeployment) rules in Cluster API you can set the `machineset.cluster.x-k8s.io/skip-preflight-checks` on a MachineDeployment and it will be 

> automatically set on the MachineSets of that MachineDeployment. You will not have to worry about setting the annotation on the new MachineSet when the 
> MachineDeployment performs a rollout.

Reviewer 2: ```suggestion
automatically set on the MachineSets of that MachineDeployment, including also the new MachineSet created when the MachineDeployment performs a rollout.
```
